src/main/java/org/testng/internal/TestNGClassFinder.java

```diff
@@ -203,10 +203,9 @@ public TestNGClassFinder(ClassInfoMap cim,
     //
     // Add all the instances we found to their respective IClasses
     //
-    for(Class c : m_instanceMap.keySet()) {
-      List<Object> instances= m_instanceMap.get(c);
-      for(Object instance : instances) {
-        IClass ic= getIClass(c);
+    for(Map.Entry<Class, List<Object>> entry : m_instanceMap.entrySet()) {
+      for(Object instance : entry.getValue()) {
+        IClass ic= getIClass(entry.getKey());
```

Comment:
Extract the line from the second loop. It will avoid to look up the same thing many times.